pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go

```diff
@@ -0,0 +1,279 @@
+	addonoperator "github.com/open-cluster-management/klusterlet-addon-controller/pkg/components/addon-operator/v1"
+	"github.com/open-cluster-management/library-go/pkg/applier"
+	"github.com/open-cluster-management/library-go/pkg/templateprocessor"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	installationNamespaceAnnotation = "addon.open-cluster-management.io/installNamespace"
+	registrationAnnotation          = "addon.open-cluster-management.io/registrations"
+	registrationConfig              = `[{"signerName":"kubernetes.io/kube-apiserver-client"}]`
+)
+
+var merger applier.Merger = func(current,
+	new *unstructured.Unstructured,
+) (
+	future *unstructured.Unstructured,
+	update bool,
+) {
+	if spec, ok := new.Object["spec"]; ok &&
+		!reflect.DeepEqual(spec, current.Object["spec"]) {
+		update = true
+		current.Object["spec"] = spec
+	}
+	if rules, ok := new.Object["rules"]; ok &&
+		!reflect.DeepEqual(rules, current.Object["rules"]) {
+		update = true
+		current.Object["rules"] = rules
+	}
+	if roleRef, ok := new.Object["roleRef"]; ok &&
+		!reflect.DeepEqual(roleRef, current.Object["roleRef"]) {
+		update = true
+		current.Object["roleRef"] = roleRef
+	}
+	if subjects, ok := new.Object["subjects"]; ok &&
+		!reflect.DeepEqual(subjects, current.Object["subjects"]) {
+		update = true
+		current.Object["subjects"] = subjects
+	}
+	return current, update
+}
+
+func createOrUpdateHubKubeConfigResources(
+	klusterletaddonconfig *agentv1.KlusterletAddonConfig,
+	r *ReconcileManagedClusterAddOn,
+	addon addons.KlusterletAddon) error {
+	componentName := addon.GetAddonName()
+
+	//Create the values for the yamls
+	config := struct {
+		ManagedClusterName      string
+		ManagedClusterNamespace string
+		ComponentName           string
+		ClusterRoleName         string
+	}{
+		ManagedClusterName:      klusterletaddonconfig.Name,
+		ManagedClusterNamespace: klusterletaddonconfig.Name,
+		ComponentName:           componentName,
+		ClusterRoleName:         addons.GetAddonClusterRolePrefix() + addon.GetManagedClusterAddOnName(),
+	}
+
+	newApplier, err := applier.NewApplier(
+		bindata.NewBindataReader(),
+		&templateprocessor.Options{},
+		r.client,
+		klusterletaddonconfig,
+		r.scheme,
+		merger,
+		&applier.Options{
+			Backoff: &wait.Backoff{
+				Steps:    1,
+				Duration: 10 * time.Millisecond,
+				Factor:   1.0,
+			},
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	err = newApplier.CreateOrUpdateInPath(
+		"resources/hub/common",
+		nil,
+		false,
+		config,
+	)
+	if err != nil {
+		return err
+	}
+	// delete old role & rolebindings created in previous releases
```

Comment:
Im not sure we really need to do this.  I would remove deletion from this PR. The concern is mainly on version skew.